staging/kos/pkg/controllers/connectionagent/connection_agent.go

```diff
@@ -0,0 +1,1124 @@
+	kosclientset "k8s.io/examples/staging/kos/pkg/client/clientset/versioned"
+	kosscheme "k8s.io/examples/staging/kos/pkg/client/clientset/versioned/scheme"
+	netifcv1a1 "k8s.io/examples/staging/kos/pkg/client/clientset/versioned/typed/network/v1alpha1"
+	kosinformers "k8s.io/examples/staging/kos/pkg/client/informers/externalversions"
+	koslisterv1a1 "k8s.io/examples/staging/kos/pkg/client/listers/network/v1alpha1"
+	netfabric "k8s.io/examples/staging/kos/pkg/networkfabric"
+	"k8s.io/examples/staging/kos/pkg/util/parse"
+)
+
+const (
+	// localAttsInformerFakeVNI is the fake VNI used to identify the informer on
+	// local NetworkAttachments. Remote NetworkAttachments informers are
+	// partitioned by VNI so they can be identified via their VNI, but the local
+	// NetworkAttachments informer is not associated to a single VNI. Pick 0
+	// because it's not a valid VNI: there's no overlapping with remote
+	// NetworkAttachments informers' VNIs.
+	localAttsInformerFakeVNI = 0
+
+	// Name of the indexer used to match pre-existing network interfaces to
+	// network attachments.
+	ifcOwnerDataIndexerName = "ifcOwnerData"
+
+	// Field names of NetworkAttachments used to build field selectors.
+	attNodeField   = "spec.node"
+	attIPv4Field   = "status.ipv4"
+	attHostIPField = "status.hostIP"
+	attVNIField    = "status.addressVNI"
+
+	// resync period for Informers caches. Set
+	// to 0 because we don't want resyncs.
+	resyncPeriod = 0
+
+	// netFabricRetryPeriod is the time we wait before retrying when a
+	// network fabric operation fails while handling pre-existing interfaces.
+	netFabricRetryPeriod = time.Second
+
+	// The HTTP port at which Prometheus metrics are served.
+	// Pick an unusual one because the host's network namespace is used.
+	// See https://github.com/prometheus/prometheus/wiki/Default-port-allocations .
+	metricsAddr = ":9294"
+
+	// The HTTP path at which Prometheus metrics are served.
+	metricsPath = "/metrics"
+
+	// The namespace and subsystem of the Prometheus metrics produced here
+	metricsNamespace = "kos"
+	metricsSubsystem = "agent"
+)
+
+// layer1VirtualNetworkState is the first layer of state associated with a
+// single relevant virtual network.
+type layer1VirtualNetworkState struct {
+	// List of names of remote NetworkAttachments in the virtual network (which
+	// implicitly determines the namespace) for whom an add notification handler
+	// has been executed but a delete notification handler has not.
+	remoteAtts map[string]struct{}
+
+	// Lister used by workers to retrieve the NetworkAttachment they're
+	// processing.
+	remoteAttsLister koslisterv1a1.NetworkAttachmentNamespaceLister
+}
+
+// layer1VirtualNetworksState is the first layer of state associated with all
+// the relevant virtual networks.
+// Its main purposes are retrieval of remote NetworkAttachments by workers and
+// deletion of network interfaces of those remote NetworkAttachments when the
+// virtual network becomes irrelevant.
```

Comment:
I think the description of the distinction between the two layers can be improved.

The responsibility for "deletion of network interfaces of those remote NetworkAttachments when the virtual network becomes irrelevant" is diffused far and wide, including both layer 2 and the work queue workers.  Even the detection of when a virtual network becomes irrelevant is in layer 2.

I would say that layers 1 and 2 are jointly responsible for synthesizing the effect of one dynamically selective informer from a dynamic collection of statically selective informers.  The distinction is that layer 1 is designed so that a notification handler can always access it promptly and briefly and layer 2 is not designed for access by notification handlers at all.

Reply:
I revamped the comments on the two layers in 588ee11 in #85 .

The new comments contain all the information you're suggesting and more details to give some context.

The fact that one layer is engineered to be accessed by notification handlers and the other is not is mentioned but is not presented as the primary distinction between the two layers, IMO that fact is not worth that much importance; I think a more useful perspective is: each layer serves different purposes and as such carries different information, the information in layer 1 happens to be available in notification handlers.

One difficulty was choosing where to put the new comments. I think there should be comments on top of the type definitions, but those are separate and some comments need to compare the two layers. I added the information that pertains to single layers on top of the relevant type definitions and the information that compares the two layers on top of the two ConnectionAgent struct fields (defined one after the other) that store the two layers.
Personally I find comments in the latter location easier to miss and wonder whether it could be worth it to introduce a general struct type with two fields, one for each layer, such as
```
type virtualNetworksState struct {
    layer1 layer1VirtualNetworksState
    layer2 layer2VirtualNetworksState
} 
```
the advantage is that we have a centralized place where to put all the information (both per-layer and general) on the layers, the disadvantage is that the expressions to access the layers' fields in the code become longer (by one level of field indirection) e.g. from:
```
ca.layer1VirtNetsState.<field-name>
```
to
```
ca.virtNetsState.layer1.<field-name>
```